analyzing: Infer global constant declarations in JS/TS

IagoAbal opened the pull request:
Even if the `const' qualifier is missing; it's enough if the variable is
resolved, it's set to a literal, and it's only assigned once.

Closes: #554



PR checklist:
- [ x] changelog is up to date

aryx commented on semgrep-core/src/analyzing/Constant_propagation.ml:
```diff
@@ -116,15 +116,19 @@ let var_stats prog : var_stats =
             incr stat.lvalue;
-            k x
+            (* We can't do `k x` otherwise the variable-declaration itself is
+             * re-visited but now interpreted as an assignment, and we would be
```
Please, add "because of v_vardef_as_assign_expr in Visitor_AST.ml".
This v_vardef_as_assign_expr is maybe not a good thing ... We should maybe do that in Generic_vs_generic
instead of in the visitor.

aryx reviewed (approved):
Looks good. Need to look for the CI errors though.

IagoAbal committed:
analyzing: Infer global constant declarations in JS/TS

Even if the `const' qualifier is missing; it's enough if the variable is
resolved, it's set to a literal, and it's only assigned once.

Closes: #554

IagoAbal commented:
There seem to be other semgrep-rules failing now... looking into it.

IagoAbal commented:
For the tests to pass here we need to update semgrep-rules, but updating semgrep-rules will make the tests fail with develop.

So I checked that all e2e tests passed except for `test_semgrep_rules_repo`: 
```
pytest tests/e2e/ -k 'not test_semgrep_rules_repo'
```
And I opened a PR to semgrep-rules with the required changes: https://github.com/returntocorp/semgrep-rules/pull/1269.
And I check that this modified Semgrep passes the tests for the modified semgrep-rules.

IagoAbal commented:
Wait, better run all QA and Unit tests too...

IagoAbal commented:
OK, unit tests pass too. QA tests, 127 pass and 17 "xfail", but those 17 "xfailures" also happen with develop so I guess it's fine?

aryx commented:
yes, xfail are ok.

aryx commented on semgrep/tests/e2e/snapshots/test_check/test_basic_rule__absolute/results.json:
```diff
@@ -56,12 +56,12 @@
               "col": 14,
               "line": 3,
-              "offset": 25
+              "offset": 62
```
Why the offset changed after this PR?

aryx replied:
Ah ok got it, because of the window.prompt() modification to the test file.
